src/riak_kv_qry_buffers.erl

```diff
@@ -0,0 +1,815 @@
+                          is_ready      = IsReady,
+                          last_accessed = LastAccessed,
+                          total_records = TotalRecords}} <- QBufs],
+    {reply, Res, State}.
+
+
+do_get_total_size(#state{total_size = TotalSize} = State) ->
+    {reply, TotalSize, State}.
+
+
+do_get_or_create_qbuf(SQL, NSubqueries,
+                      OrigDDL, Options,
+                      #state{qbufs            = QBufs0,
+                             soft_watermark   = SoftWMark,
+                             root_path        = RootPath,
+                             total_size       = TotalSize,
+                             qbuf_expire_msec = DefaultQBufExpireMsec} = State0) ->
+    case maybe_ensure_qref(SQL, QBufs0) of
+        {ok, {existing, QBufRef}} ->
+            lager:info("reusing existing query buffer ~p for ~p", [QBufRef, SQL]),
+            State9 = touch_qbuf(QBufRef, State0),
+            {reply, {ok, {existing, QBufRef}}, State9};
+        {ok, {new, QBufRef}} ->
+            if TotalSize > SoftWMark ->
+                    {reply, {error, total_qbuf_size_limit_reached}, State0};
+               el/=se ->
+                    DDL = ?DDL{table = Table} = sql_to_ddl(OrigDDL, SQL),
+                    lager:info("creating new query buffer ~p (ref ~p) for ~p", [Table, QBufRef, SQL]),
+                    case riak_kv_qry_buffers_ldb:new_table(Table, RootPath) of
+                        {ok, LdbRef} ->
+                            QBuf = #qbuf{orig_qry      = SQL,
+                                         ddl           = DDL,
+                                         mother_table  = riak_kv_ts_util:queried_table(SQL),
+                                         chunks_need   = NSubqueries,
+                                         ldb_ref       = LdbRef,
+                                         expire_msec   = proplists:get_value(
+                                                           expiry_msec, Options, DefaultQBufExpireMsec),
+                                         last_accessed = os:timestamp(),
+                                         options       = Options,
+                                         key_field_positions = get_lk_field_positions(DDL)},
+                            QBufs = QBufs0 ++ [{QBufRef, QBuf}],
+                            State = State0#state{qbufs = QBufs,
+                                                 total_size = compute_total_qbuf_size(QBufs)},
+                            {reply, {ok, {new, QBufRef}}, State};
+                        {error, Reason} ->
+                            {reply, {error, Reason}, State0}
+                    end
+            end;
+        {error, Reason} ->
+            {reply, {error, Reason}, State0}
+    end.
+
+maybe_ensure_qref(?SQL_SELECT{allow_qbuf_reuse = true} = SQL, QBufs) ->
+    ensure_qref(SQL, QBufs);
+maybe_ensure_qref(?SQL_SELECT{allow_qbuf_reuse = false}, _QBufs) ->
+    AlwaysUniqueRef = crypto:rand_bytes(8),
```

Comment:
I need it to be a binary, the same size as that generated by `make_qref` (if only for it to be of the same type from dialyzer's POV).